frame/evm/src/runner/mod.rs

```diff
@@ -0,0 +1,82 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod stack;
+pub mod native;
+
+use sp_std::vec::Vec;
+#[cfg(feature = "std")]
+use serde::{Serialize, Deserialize};
+use codec::{Encode, Decode};
+use sp_core::{H160, U256, H256};
+use evm::ExitReason;
+use crate::Trait;
+
+pub use evm::backend::{Basic as Account, Log};
+
+#[derive(Clone, Eq, PartialEq, Encode, Decode, Default)]
+#[cfg_attr(feature = "std", derive(Debug, Serialize, Deserialize))]
+/// External input from the transaction.
+pub struct Vicinity {
+	/// Current transaction gas price.
+	pub gas_price: U256,
+	/// Origin of the transaction.
+	pub origin: H160,
+}
+
+pub struct ExecutionInfo<T> {
+	pub exit_reason: ExitReason,
+	pub value: T,
+	pub used_gas: U256,
+	pub logs: Vec<Log>,
+}
+
+pub type CallInfo = ExecutionInfo<Vec<u8>>;
+pub type CreateInfo = ExecutionInfo<H160>;
+
+pub trait Runner<T: Trait> {
+	type Error: Into<sp_runtime::DispatchError>;
+
+	fn call(
+		source: H160,
+		target: H160,
+		input: Vec<u8>,
+		value: U256,
+		gas_limit: u32,
+		gas_price: U256,
+		nonce: Option<U256>,
```

Comment:
@sorpaas I think we need `apply_state` argument so we can use it for `eth_call` and `eth_estimateGas`

Reply:
We don't, actually!

What `eth_call` and `eth_estimateGas` do in Frontier is to spawn a new runtime instance based on a particular block, run functions, and retrieve the results. The runtime instance is then discarded, hence the state changes as well.

Previously we had `apply_state` because it brings slight performance improvements -- state applications happen at the end of the block, and `eth_call` and `eth_estimateGas` do not need them, so with the flag setting to false we can early return. Right now, however, it is really hard to support this apply state flag in native runner, because we instantly apply changes as the EVM runs. So for a generic interface we had to remove it.

Reply:
Hi, @sorpaas, I  have many puzzles about `apply_state` usage here. Especially about these comments:

> Previously we had apply_state because it brings slight performance improvements -- state applications happen at the end of the block, and eth_call and eth_estimateGas do not need them, so with the flag setting to false we can early return.

The following is what I understand about this:
 
```rs
fn execute_evm<F, R>(
	......
	if apply_state {
		backend.apply(values, logs_data, true);
	}
}
```

In evm-pallet, all `execute_call`, `execute_create` finally use `execute_evm` to start evm `StackExecutor` running
and get result. At the end of `execute_evm`, if `apply_state` is true, we will change the `AccountCodes`, `AccountStorages` and so on. If `apply_state` is false,  we skip the evm-store values modifactions and return early.  

While we use `eth_call` or  `eth_estimateGas`, we set the `apply_state` flag with false to avoid modification storage stored in the pallet. This makes sense to me.

Also, I have noticed this comment:

> The runtime instance is then discarded, hence the state changes as well.`

I am confused about `state changes` here. Could someone do a more clear explanation about this?

Thanks.